### PR TITLE
Feature: code coverage test

### PR DESCRIPTION
**Pull request by marehr**
This introduces code coverage testing:

* `cd /seqan3 && mkdir -p build/coverage  && cmake ../../test/coverage -DCMAKE_BUILD_TYPE=Debug`
* `make` or `make coverage` will generate coverage reports
* `make coverage_html` will generate a graphical report

This CMakeScript uses under-the-hood `g++ --coverage` and `lcov` (which uses `gcov`).

`lcov` is used to merge coverage reports and also to filter out `/usr/*` and some other directories.

The script was adopted from https://github.com/bilke/cmake-modules/blob/master/CodeCoverage.cmake (author was included into the copyright)

Disclaimer: g++-8 seems to have a broken `lcov`, reports generated with g++-7 seems to be fine.

partially resolves (#167)

Resources:
* http://ltp.sourceforge.net/coverage/lcov.php
* https://codecov.io/gh/codecov/example-cpp11-cmake/src/master/src/complex_main.cpp
* https://github.com/codecov/example-cpp11-cmake
* https://docs.codecov.io/docs
* https://clang.llvm.org/docs/SourceBasedCodeCoverage.html

------

blocked by #222: CMake changes

### CODE REVIEW

**Comment by rrahn** [comment]
@marehr at the first glance, this PR does more than just adding coverage tests, but involves a lot of restructuring. Please separate the structural changes from the coverage feature in another PR.

File: test/seqan3-test.cmake [hunk 1]
```diff
@@ -48,22 +48,40 @@ include (FindPackageMessage)
...
+# see cmake bug https://gitlab.kitware.com/cmake/cmake/issues/15052
+file(MAKE_DIRECTORY ${SEQAN3_BENCHMARK_CLONE_DIR}/include/)
+file(MAKE_DIRECTORY ${SEQAN3_TEST_CLONE_DIR}/googletest/include/)
+
+# seqan3::test exposes a base set of required flags, includes, definitions and
+# libraries qhich are in common for **all** seqan3 tests
+add_library (seqan3::test INTERFACE IMPORTED)
+set_property (TARGET seqan3::test APPEND PROPERTY INTERFACE_COMPILE_OPTIONS "-pedantic"  "-Wall" "-Wextra" "-Werror")
+set_property (TARGET seqan3::test APPEND PROPERTY INTERFACE_LINK_LIBRARIES "seqan3::seqan3" "pthread")
+set_property (TARGET seqan3::test APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${SEQAN3_CLONE_DIR}/test/include/")
```
Comment (rrahn) [inline test/seqan3-test.cmake]: Mhmm I am a little bit sceptical of putting this into the global seqan3-test configuration. 
What would be the benefit of knowing these targets in all build projects? 
There is no reuse within the target specific builds, is there?

File: test/seqan3-test.cmake [hunk 2]
```diff
@@ -97,88 +115,92 @@ endmacro ()
...
+    unset(gtest_path)
 endmacro ()
 
-macro (add_subdirectories)
+macro (add_subdirectories_of directory)
     file (GLOB ENTRIES
-          RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}
-          ${CMAKE_CURRENT_SOURCE_DIR}/[!.]*)
+          RELATIVE ${directory}
+      ${directory}/[!.]*)
```
Comment (rrahn) [inline test/seqan3-test.cmake]: Indentation

File: test/coverage/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    message(FATAL_ERROR "lcov not found! Aborting...")
+endif ()
+
+set (COVERAGE_EXCLUDES
+    "'/usr/*'"
+    "'${SEQAN3_CLONE_DIR}/submodules/*'"
+    "'${SEQAN3_CLONE_DIR}/test/unit/*'"
+    "'${PROJECT_BINARY_DIR}/vendor/*'"
+)
+set_property (GLOBAL PROPERTY COVERAGE_ALL_TESTS "")
```
Comment (rrahn) [inline test/coverage/CMakeLists.txt]: ??? Can you add a comment on how to use this property.
Also the naming does not fit the naming conventions for the projects. I would rather use `TEST_COVERAGE_ALL_TARGETS`.

**Comment by marehr** [comment]
@rrahn I wrote that this PR is blocked by #222 (this PR includes the same changes + the code coverage ones)

File: test/coverage/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+endmacro ()
+
+seqan3_require_ccache ()
+seqan3_require_test ()
+
+# add all unit tests
+add_subdirectories_of ("${CMAKE_CURRENT_SOURCE_DIR}/../unit")
+
+# add collected test cases as dependency
+get_property(COVERAGE_ALL_TESTS GLOBAL PROPERTY COVERAGE_ALL_TESTS)
```
[the remark this answers is not in the document]
Comment (marehr) [reply test/coverage/CMakeLists.txt]: Variables declared with `set()` can't be modified by `CMakeLists.txt` which are included by `add_subdirectory`. They create a new scope and all changes to a `set()` variable will be discharged.

In this case we want to define a `GLOBAL` variable which is accessible in all `CMakeLists.txt`.

File: test/coverage/CMakeLists.txt (hunk 3 above)
Comment (marehr) [reply test/coverage/CMakeLists.txt]: Usage is here: `set_property(GLOBAL APPEND PROPERTY COVERAGE_ALL_TESTS ${target})`

File: test/seqan3-test.cmake (hunk 2 above)
Comment (marehr) [reply test/seqan3-test.cmake]: good catch

File: test/coverage/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    # $target_cpp = "pod_tuple_test.cpp"
+    #   * will register the global TARGET name "pod_tuple_test" and
+    #   * will register the test case name "core_pod_tuple" if
+    #     pod_tuple_test.cpp is in test/core/
+    #
+    # NOTE(marehr): ".+/test/" REGEX is greedy, that means
+    # /test/test/test/hello_test.cpp will result in an empty `test_path`
+    string (REGEX REPLACE "_test.cpp$" "" target_name ${target_cpp})
+    string (REGEX REPLACE ".+/test/" "" test_path ${CMAKE_CURRENT_LIST_DIR})
+    set (target "${target_name}_test")
```
Comment (rrahn) [inline test/coverage/CMakeLists.txt]: maybe call it `_coverage_test`

File: test/seqan3-test.cmake [hunk 6]
```diff
@@ -48,22 +48,40 @@ include (FindPackageMessage)
...
-set (SEQAN3_TEST_LIBRARIES "")
-list (APPEND SEQAN3_TEST_INCLUDE_DIRS "${SEQAN3_TEST_CLONE_DIR}/googletest/include/")
-list (APPEND SEQAN3_TEST_INCLUDE_DIRS "${SEQAN3_CLONE_DIR}/test/include/")
+# needed for add_library (seqan3::test::* INTERFACE IMPORTED)
+# see cmake bug https://gitlab.kitware.com/cmake/cmake/issues/15052
+file(MAKE_DIRECTORY ${SEQAN3_BENCHMARK_CLONE_DIR}/include/)
+file(MAKE_DIRECTORY ${SEQAN3_TEST_CLONE_DIR}/googletest/include/)
+
+# seqan3::test exposes a base set of required flags, includes, definitions and
+# libraries qhich are in common for **all** seqan3 tests
```
[the remark this answers is not in the document]
Comment (marehr) [reply test/seqan3-test.cmake]: will be resolved in #222

**Comment by marehr** [comment]
@rrahn I addressed your issues; Can you re-review?

**Comment by h-2** [comment]
Is this tied into the new jenkins-pipeline thing? if not, can you speak with @rrahn to get it in?

**Comment by rrahn** [comment]
@marehr does the `${PROJECT_BINARY_DIR}/seqan3_coverage` contains the statistics? Is this printed somewhere in the console during execution. We won't access the html file from within the travis/jenkins build. But at least in jenkins we could access the workspace. Have you looked into the coverage feature of ctest? This will also publish a coverage report on cdash.

**Comment by marehr** [comment]
@rrahn 

> Have you looked into the coverage feature of ctest?

No I haven't.

> Is this printed somewhere in the console during execution. We won't access the html file from within the travis/jenkins build.

As you can see in https://travis-ci.org/seqan/seqan3/jobs/405798259 it will print out:

> Deleted 242 files
Writing data to /home/travis/build/seqan/seqan3-build/seqan3_coverage
Summary coverage rate:
  lines......: 76.6% (1904 of 2485 lines)
  functions..: 77.4% (3131 of 4043 functions)
  branches...: no data found

I thought we use https://codecov.io/, this PR uses https://github.com/codecov/example-cpp11-cmake as base to create the report.

**Comment by marehr** [comment]
report looks like this: https://codecov.io/gh/seqan/seqan3/pull/271

File: test/coverage/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+endmacro ()
+
+seqan3_require_ccache ()
+seqan3_require_test ()
+
+# add all unit tests
+add_subdirectories_of ("${CMAKE_CURRENT_SOURCE_DIR}/../unit")
+
+# add collected test cases as dependency
+get_property(TEST_COVERAGE_ALL_TESTS GLOBAL PROPERTY TEST_COVERAGE_ALL_TESTS)
```
Comment (rrahn) [inline test/coverage/CMakeLists.txt]: Give the target of the command a distinguishable name from the global property name. Maybe just add a underscore at the beginning.

**Pull request merged**